OutboxSample/Application/DataAccess/IRepository.cs

```diff
@@ -0,0 +1,68 @@
+﻿namespace OutboxSample.Application.DataAccess;
+
+public interface IRepository
+{
+}
+
+public abstract record QueryResult
+{
+    public static Success<TData> OfSuccess<TData>(TData data, DataStore.ItemMetadata metadata) => new(data, metadata);
+
+    public sealed record Success<TData>(TData Data, DataStore.ItemMetadata Metadata) : QueryResult<TData>, Success;
+
+    public interface Success { }
+
+    public interface Failure { }
+}
+
+public abstract record QueryResult<TExpectedData> : QueryResult
+{
+    public static class OfFailure
+    {
+        public static Failure AlreadyExists(string message = "Already exists.") => new Failure.AlreadyExists(message);
+
+        public static Failure NotFound(string message = "Not found.") => new Failure.NotFound(message);
+
+        public static Failure ConcurrencyConflict(string message = "Concurrency conflict.") => new Failure.ConcurrencyConflict(message);
+    }
+
+    new public abstract record Failure(Failure.Description WhatHappened) : QueryResult<TExpectedData>, QueryResult.Failure
+    {
+        public sealed record AlreadyExists(string Message) : Failure(Message, ErrorCode.AlreadyExists);
+
+        public sealed record NotFound(string Message) : Failure(Message, ErrorCode.NotFound);
+
+        public sealed record ConcurrencyConflict(string Message) : Failure(Message, ErrorCode.ConcurrencyConflict);
+
+        protected Failure(string message, ErrorCode errorCode) : this(new Description(message, errorCode))
+        { }
+
+        public readonly record struct Description(string Text, ErrorCode ErrorCode);
+
+        public enum ErrorCode
+        {
+            Undefined = 0,
+            // Unexpected = 1,
+            NotFound = 2,
+            AlreadyExists = 3,
+            ConcurrencyConflict = 4,
+        }
+    }
+}
+
+public static class DataStore
+{
+    // NOTE: need to watch out for how much data it carries around as after some threshold it might make sense to conver it to ref-type record
```

Comment:
That note is ominous, maybe this should be a `class` from the start?

Reply:
I'll better remove the note ;) I don't think copying a value type like that will ever be a real problem.